src/sync/partial_sync.cpp

```diff
@@ -56,6 +83,153 @@ void update_schema(Group& group, Property matches_property)
+    }
+
+private:
+    void create_thread()
+    {
+        if (m_thread.joinable())
+            m_thread.join();
+
+        m_thread = std::thread([this] {
+            std::vector<std::function<void()>> queue;
+
+            std::unique_lock<std::mutex> lock(m_mutex);
+            while (!m_stopping &&
+                   m_cv.wait_for(lock, std::chrono::milliseconds(500),
+                                 [&] { return !m_queue.empty() || m_stopping; })) {
+
+                swap(queue, m_queue);
+
+                lock.unlock();
+                for (auto& f : queue)
+                    f();
+                queue.clear();
+                lock.lock();
+            }
+
+            m_stopped = true;
+        });
+        m_stopped = false;
+    }
+
+    std::mutex m_mutex;
+    std::condition_variable m_cv;
+    std::vector<std::function<void()>> m_queue;
+    std::thread m_thread;
+    bool m_stopping = false;
+    bool m_stopped = true;
+};
+
+static auto& work_queue = *new WorkQueue;
+
+void async_register_query(Realm& realm, std::string object_type, std::string query, std::string name,
+                          std::function<void(std::exception_ptr)> callback)
+{
+    work_queue.enqueue([object_type=std::move(object_type), query=std::move(query), name=std::move(name),
+                        callback=std::move(callback), config=realm.config()] {
+        auto realm = Realm::get_shared_realm(config);
+        realm->begin_transaction();
+        auto cleanup = util::make_scope_exit([&]() noexcept {
+            if (realm->is_in_transaction())
+                realm->cancel_transaction();
+        });
+
+        auto matches_property = std::string(object_type) + "_matches";
+
+        update_schema(realm->read_group(), Property(matches_property, PropertyType::Object|PropertyType::Array, object_type));
+        ObjectSchema result_sets_schema(realm->read_group(), result_sets_type_name);
+
+        try {
+            if (!validate_existing_subscription(realm, name, query,
+                                                matches_property, result_sets_schema)) {
+                CppContext context;
+                auto object = Object::create<util::Any>(context, realm, result_sets_schema,
+                                                        AnyDict{
+                                                            {"matches_property", matches_property},
+                                                            {"name", name},
+                                                            {"query", query},
+                                                            {"status", int64_t(0)},
+                                                            {"error_message", std::string()},
+                                                            {"query_parse_counter", int64_t(0)},
+                                                        }, false);
+            }
+        } catch (...) {
+            callback(std::current_exception());
```

Comment:
This callback seems to happen outside the notifier framework? In order for Java to couple these errors + notifications from the CollectionNotifier, we need to be able to be able to reason about the latest state of both.

Reply:
After looking at the Subscription `struct the `finished_subscribing` method seems to handle this?

Reply:
The callback that is invoked here is the one within `subscribe` that routes the error, if any, to `Notifier::finished_subscribing`. The `Notifier` handles kicking things over to the originating thread by hooking in to the collection notification subsystem.